File: ch03/ex3.95.jl
## 1. load package
include("../utils.jl")
  using GLMakie,CSV,DataFrames,Random,StatsBase,PrettyTables,Pipe

##  2. load data
  #= 
    feature :["Butter", "Filling", "Bread", "Ants", "Order"]
  =#
  desc=Lock5Table(281,"SandwichAnts","3.95 Ants on a Sandwich",["Butter", "Filling", "Bread", "Ants", "Order"])
  df=@pipe load_csv(desc.name)
  peanbutter_ants=filter(row -> row.Filling == "Peanut Butter", df)
  #= 
    8×5 DataFrame
    Row │ Butter   Filling        Bread       Ants   Order 
        │ String3  String15       String15    Int64  Int64 
    ─────┼──────────────────────────────────────────────────
    1 │ no       Peanut Butter  Rye            43     26
    2 │ no       Peanut Butter  Wholemeal      59     35
    3 │ no       Peanut Butter  Multigrain     22     36
    4 │ no       Peanut Butter  White          25     34
    5 │ no       Peanut Butter  Rye            36     31
    6 │ no       Peanut Butter  Wholemeal      47     38
    7 │ no       Peanut Butter  Multigrain     19     22
    8 │ no       Peanut Butter  White          21     16
  =#

## 3. Bootstrap sampling workflow
### 3.1  data
    data=peanbutter_ants.Ants
### 3.2 mean and std
    @info "mean_std"=>mean_and_std(data)
    # [ Info: "mean_std" => (34.0, 14.628738838327793)

### 3.3  define function 
    cil = 0.95;n_boot=5000
    """
    bs(data::Vector)
    bootstraping 包装方法
    """
    bs(data::Vector) = bootstrap(mean,data, BasicSampling(n_boot))
### 3.4  bootstrap samping
     bs1=bs(data)
     #= 
       Bootstrap Sampling
        Estimates:
            Var │ Estimate  Bias     StdError
                │ Float64   Float64  Float64
            ─────┼─────────────────────────────
            1 │     34.0  0.10995   4.81507
        Sampling: BasicSampling
        Samples:  5000
        Data:     Vector{Int64}: { 8 }
     =#

### 3.5 confidence
    bci1 = confint(bs1, BasicConfInt(cil))
    # ((34.0, 24.125, 42.75),)

File: marimo/ch03/ex3.95.py
import marimo

__generated_with = "0.24.0"
app = marimo.App(width="medium")


@app.cell
def _():
    import marimo as mo
    import sys
    from pathlib import Path
    import numpy as np
    import polars as pl
    import matplotlib.pyplot as plt
    from scipy import stats
    from great_tables import GT
    _p = Path(mo.notebook_location()).resolve().parent
    while _p != _p.parent and not (_p / "lock5.py").exists():
        _p = _p.parent
    if str(_p) not in sys.path:
        sys.path.insert(0, str(_p))
    import lock5
    return mo, sys, Path, np, pl, plt, stats, GT, lock5


@app.cell
def _(mo):
    mo.md(r"""# Ex 3.95 — ants on a sandwich (bootstrap)

Port of `UnitB/ch03/ex3.95.jl`. Bootstrap the mean number of ants for the
"Peanut Butter" filling (5000 samples) and compute a basic 95% CI. Bootstrap
results are random, so values differ from the Julia run.""")
    return


@app.cell
def _(lock5, pl):
    desc = lock5.Lock5Table(281, "SandwichAnts", "3.95 Ants on a Sandwich",
                            ["Butter", "Filling", "Bread", "Ants", "Order"])
    df = lock5.load_csv(desc.name)
    peanbutter_ants = df.filter(pl.col("Filling") == "Peanut Butter")
    data = peanbutter_ants["Ants"].to_numpy().astype(float)
    return data, desc, df, peanbutter_ants


@app.cell
def _(data, lock5, mo, np, pl):
    mn, sd = lock5.mean_and_std(data)
    boots = lock5.boot_sampling(data, n=5000, method=np.mean, seed=1)
    estimate = float(data.mean())
    bci = lock5.basic_confint(boots, estimate, 0.95)
    res_df = pl.DataFrame({
        "item": ["mean", "std (sample)", "bias", "std error",
                 "basic 95% CI lower", "basic 95% CI upper"],
        "value": [round(mn, 4), round(sd, 4), round(float(boots.mean() - estimate), 4),
                  round(float(boots.std(ddof=1)), 4), round(bci[1], 4), round(bci[2], 4)],
    })
    mo.Html(lock5.gt_table(res_df, title="Peanut Butter ants bootstrap (5000 samples)").as_raw_html())
    return bci, boots, estimate, res_df


if __name__ == "__main__":
    app.run()
